src/SubscriptionStore.h

```diff
@@ -29,40 +33,194 @@ class SubscriptionStore
+                "Attempt to remove subscription for connection {} / channel {}, "
+                "but no subscriptions for this connection exist",
+                connection->GetHostname(),
+                channelId);
+            success = false;
+        }
+        else
+        {
+            auto& subs = subscriptionsByConnection[connection];
+            int numSubsErased = 0;
+            for (auto it = subs.begin(); it != subs.end();)
+            {
+                auto& sub = *it;
+                if (sub->ChannelId == channelId)
+                {
+                    it = subs.erase(it);
+                    ++numSubsErased;
+                }
+                else
+                {
+                    ++it;
+                }
+            }
+            if (numSubsErased == 0)
+            {
+                spdlog::error(
+                    "Attempt to remove non-existant subscription for connection {} to channel {}",
+                    connection->GetHostname(),
+                    channelId);
+                success = false;
+            }
+        }
+
+        if (subscriptionsByChannel.count(channelId) <= 0)
+        {
+            spdlog::error(
+                "Attempt to remove subscription for connection {} / channel {}, "
+                "but no subscriptions to this channel exist",
+                connection->GetHostname(),
+                channelId);
+            success = false;
+        }
+        else
+        {
+            auto& subs = subscriptionsByChannel[channelId];
+            int numSubsErased = 0;
+            for (auto it = subs.begin(); it != subs.end();)
+            {
+                auto& sub = *it;
+                if (sub->SubscribedConnection == connection)
+                {
+                    it = subs.erase(it);
+                    numSubsErased++;
+                }
+                else
+                {
+                    ++it;
+                }
+            }
+
+            if (numSubsErased == 0)
+            {
+                spdlog::error(
+                    "Attempt to remove non-existant subscription on channel {} for connection {}",
+                    connection->GetHostname(),
+                    channelId);
+                success = false;
+            }
+        }
+        return success;
+    }
 
     /**
      * @brief Get the list of channel subscriptions that exist for a connection
      * @param connection connection to fetch subscribed channels for
      * @return std::set<ftl_channel_id_t> set of channels this connection is subscribed to
      */
-    std::set<ftl_channel_id_t> GetSubscribedChannels(std::shared_ptr<IConnection> connection);
+    std::vector<ChannelSubscription<TConnection>> GetSubscriptions(std::shared_ptr<TConnection> connection)
+    {
+        std::vector<ChannelSubscription<TConnection>> returnVal;
+        std::lock_guard<std::mutex> lock(subscriptionsStoreMutex);
+        if (subscriptionsByConnection.contains(connection))
```

Comment:
No need to change the code right now, trying to remember how funky C++ constructors are.

Is it possible simplify this by either using the vector range constructor:
```
const auto& subscriptions = subscriptionsByConnection[connection]
std::vector<ChannelSubscription<TConnection>> returnVal(subscriptions.begin(), subscriptions.end()); 
```

Or even simpler, make this function return a set and use the copy constructor?
```
std::map<ChannelSubscription<TConnection>> subscriptions(subscriptionsByConnection[connection]); 
```

I'm not 100% sure but my reading of the copy constructor for `map` is that it is linear time complexity if the container being copied is already in the expected sort order (which it always would be in this case), so the performance isn't worse and it's much simpler. My C++ is too rusty but probably the copy and return could be done on a single line.

(not that perf matters in this context much, if it did I think the solution would be to expose the iterators for the subscription set directly to avoid the copy when just looping. kinda annoying to have to expose multiple methods to iterate though)

Reply:
Yeah, this was just a hasty approach to rid of the `shared_ptr` and copy - I'm not sure if the iterator range constructor is smart enough to dereference those and copy for me? Or the map constructor? I can experiment.

Reply:
Tried this out, but sadly it looks like I have to dereference the shared_ptr myself, so I need a loop unless there's some other approach I don't know of.